{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "624a4ea0-1793-4432-ae27-649da9adf377",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: AnVIL\n",
      "\n",
      "Loading required package: dplyr\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: MultiAssayExperiment\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(RColorBrewer)\n",
    "library(ggpubr)\n",
    "library(RColorBrewer)\n",
    "library(ggplot2)\n",
    "library(ggpubr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c0414da-e9c6-44e5-a5d9-572318f71987",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Differences in risk scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e70d13bf-f04e-4c34-a7ac-ee043de4aafa",
   "metadata": {},
   "outputs": [],
   "source": [
    "xxx = fread(\"TCGA.COAD.sampleMap_COAD_clinicalMatrix\")\n",
    "xxx$MSI_ = ifelse(xxx$MSI ==\"Indeterminate\"|xxx$MSI ==\"\",NA,xxx$MSI)\n",
    "xxx$MSI  = ifelse( xx$MSI_updated_Oct62011 !=\"\",xx$MSI_updated_Oct62011,xxx$CDE_ID_3226963)\n",
    "xxx$MSI_ = ifelse(xxx$MSI ==\"Indeterminate\"|xxx$MSI ==\"\",NA,xxx$MSI)\n",
    "xxx = na.omit(xxx)\n",
    "save(coad_msi,file = \"coad_msi.rdata\")\n",
    "load(\"msi_risk.rdata\")\n",
    "x = as.data.frame(msi_risk[order(msi_risk$riskScore,decreasing = F),])\n",
    "x = x[-(grep(\"MSS\",head(x,50)[,2])),]\n",
    "\n",
    "mypalette <- colorRampPalette(brewer.pal(8,\"Set1\"))\n",
    "my_comparisons <- list( c(\"MSI-H\", \"MSI-L\"), c(\"MSI-H\", \"MSS\"), c(\"MSI-L\", \"MSS\") ) \n",
    "g3 = ggplot(x,aes(MSI,riskScore,fill = MSI)) + \n",
    "  geom_boxplot(outlier.shape = 21,color = \"black\") + \n",
    "  theme_bw() + \n",
    "  labs(x = \"\", y = \"riskScore\") +\n",
    "  theme(legend.position = \"top\") + \n",
    "  scale_fill_manual(values = mypalette(22)[c(6,1,2)])+ \n",
    "  stat_compare_means(comparisons = my_comparisons)+   # Add global p-value\n",
    "theme(text = element_text(size = 20))\n",
    "g3\n",
    "ggsave(\"MSI_3.pdf\",g3,width = 5,height =5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69011af2-f7dd-4b0d-b075-5deab1d8007a",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Appling KM analysis in group of MSI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06488f9a-c4b4-45a8-a8b9-f10ba2dced1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "load(\"../../Model_lasso_cox/模型构建/芯片建模_TCGA验证/OKModel_risk.rdata\")\n",
    "TCGA_risk_ = TCGA_risk[,1:10]\n",
    "TCGA_risk_$ID = substr(TCGA_risk_$ID,1,15)\n",
    "x__ = merge(x,TCGA_risk_,by.x = \"ID\",by.y =\"ID\")\n",
    "x__$MSI_risk = ifelse(x__$MSI == \"MSS\" & x__$risk == \"high\",\"MSS+High risk\",\n",
    "                     ifelse(x__$MSI == \"MSS\" & x__$risk == \"low\",\"MSS+Low risk\",\n",
    "                           ifelse(x__$MSI == \"MSI-H\" & x__$risk == \"low\",\"MSI-H+Low risk\",\n",
    "                                 ifelse(x__$MSI == \"MSI-H\"& x__$risk == \"high\",\"MSI-H+High risk\",\n",
    "                                       ifelse(x__$MSI == \"MSI-L\"& x__$risk == \"high\",\"MSI-L+High risk\",\"MSI-L+Low risk\")))))\n",
    "library(survival)\n",
    "library(survminer)\n",
    "fit <- survfit(Surv(OS_time, Status) ~x__[,\"MSI_risk\"] , data = x__)\n",
    "g = ggsurvplot(fit,\n",
    "          pval = TRUE, #conf.int = TRUE,\n",
    "          #risk.table = TRUE, # 添加风险表\n",
    "          risk.table.col = \"strata\", # 根据分层更改风险表颜色\n",
    "          #linetype = \"strata\", # 根据分层更改线型\n",
    "          surv.median.line = \"hv\", # 同时显示垂直和水平参考线\n",
    "          #ggtheme = theme_bw(), # 更改ggplot2的主题\n",
    "          palette =mypalette(22)[c(1:6)],\n",
    "               legend = c(0.9,0.8),\n",
    "               legend.title = \"\",\n",
    "          legend.labs = names(table(x__[,paste0(\"MSI_risk\")])),ggtheme = theme_minimal(base_size = 15))\n",
    "g\n",
    "pdf(\"MSI_KM.pdf\",width = 5,height = 5)\n",
    "g\n",
    "dev.off()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc6746bc-2a9a-4c0b-919f-11405d021c50",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Percentage stacking chart of MSI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "ad2d66c2-ab84-43b2-a0b4-bc87e7fd2ee8",
   "metadata": {},
   "outputs": [],
   "source": [
    "MSI_KM = x__\n",
    "MSI_boxplot = x\n",
    "save(MSI_KM,MSI_boxplot,file = \"MSI_plot.rdata\")\n",
    "ggplot(myresults_pct,aes(x=risk,y=pct,fill=MSI)) +theme_bw() + \n",
    "geom_col(width = 0.6)+\n",
    "scale_y_continuous(label = scales::percent)+scale_fill_manual(values = mypalette(22)[c(6,1,2)])+t\n",
    "heme( text = element_text(size = 15)))\n",
    "library(cBioPortalData)\n",
    "cbio <- cBioPortal()\n",
    "studies = getStudies(cbio)\n",
    "head(studies$studyId)\n",
    "id = \"ucec_tcga_pan_can_atlas_2018\"\n",
    "clinical = clinicalData(cbio, id)\n",
    "colnames(clinical)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8079f94e-962d-4c12-9542-a791e547dd9d",
   "metadata": {},
   "source": [
    "### TMB analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bfd5b9d-2453-49ae-a467-25e7268dcef7",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(TCGAbiolinks)\n",
    "library(tidyverse)\n",
    "library(maftools)\n",
    "library(data.table)\n",
    "query <- GDCquery(\n",
    "    project = \"TCGA-READ\", \n",
    "    data.category = \"Simple Nucleotide Variation\",\n",
    "    data.type = \"Masked Somatic Mutation\",\n",
    "    access = \"open\"\n",
    ")\n",
    "GDCdownload(query)\n",
    "GDCprepare(query, save = T,save.filename = \"TCGA-READ_SNP.Rdata\")\n",
    "load(file = \"TCGA-COAD_SNP.Rdata\")\n",
    "maf.read<- data\n",
    "x = fread(\"mc3.v0.2.8.PUBLIC.maf.gz\");x[1:4,1:4]\n",
    "maf <- read.maf(maf.read)\n",
    "x = tmb(maf = maf);head(x)\n",
    "tmb = data.frame(ID =gsub(\"-\",\".\" ,substr(x$Tumor_Sample_Barcode,1,16)),TMB = x$total_perMB,TMBlog = x$total_perMB_log)\n",
    "load(\"TCGA_risk.rdata\")\n",
    "risk_tmb = merge(tmb,risk,by= \"ID\")\n",
    "mypalette <- colorRampPalette(brewer.pal(8,\"Set1\"))\n",
    "my_comparisons = list(c(\"high\",\"low\"))\n",
    "g3 = ggplot(x,aes(risk,TMBlog,fill = risk)) + \n",
    "  geom_boxplot(outlier.shape = 21,color = \"black\") + \n",
    "  theme_bw() + \n",
    "  labs(x = \"\", y = \"log(TMB)\") +\n",
    "  theme(legend.position = \"top\") + \n",
    "  scale_fill_manual(values = mypalette(22)[c(1,6)])+ \n",
    "    stat_compare_means(comparisons = my_comparisons)+# 调整其他文本的字体大小\n",
    "  theme( text = element_text(size = 20))\n",
    "#stat_compare_means(aes(group = risk,label = ..p.signif..),method = \"wilcox.test\",label.x = 1.5,na.rm = T,)\n",
    "ggsave(\"TBM_TCGA.pdf\",g3,width = 5,height =5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c88f034f-e60e-45b1-8354-b84df4bd3402",
   "metadata": {},
   "source": [
    "### KM analysis in TMB group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "25d7c1e6-1036-4d0f-aecd-039040a74d3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(survival)\n",
    "library(survminer)\n",
    "load(\"../../Model_lasso_cox/模型构建/芯片建模_TCGA验证/OKModel_risk.rdata\")\n",
    "TCGA_risk_ = TCGA_risk[,1:10]\n",
    "x__ = merge(x,TCGA_risk_,by.x = \"ID\",by.y =\"ID_\")\n",
    "x__$TMB_group_risk = ifelse(x__$TMBlog > median(x__$TMBlog)&x__$risk == \"high\",\"High TMB+High risk\",\n",
    "                            ifelse(x__$TMBlog > median(x__$TMBlog)&x__$risk == \"low\",\"High TMB+Low risk\",\n",
    "                                  ifelse(x__$TMBlog < median(x__$TMBlog)&x__$risk == \"low\",\"Low TMB+Low risk\",\"Low TMB+High risk\")))\n",
    "x___ = x__[x__$OS_time<5,]\n",
    "fit <- survfit(Surv(OS_time, Status) ~x__[,\"TMB_group_risk\"] , data = x__)\n",
    "g = ggsurvplot(fit,\n",
    "          pval = TRUE, #conf.int = TRUE,\n",
    "          #risk.table = TRUE, # 添加风险表\n",
    "          risk.table.col = \"strata\", # 根据分层更改风险表颜色\n",
    "          #linetype = \"strata\", # 根据分层更改线型\n",
    "          surv.median.line = \"hv\", # 同时显示垂直和水平参考线\n",
    "          #ggtheme = theme_bw(), # 更改ggplot2的主题\n",
    "          palette =mypalette(22)[c(1:4)],\n",
    "               legend = c(0.7,0.8),\n",
    "               legend.title = \"\",\n",
    "          legend.labs = names(table(x__[,paste0(\"TMB_group_risk\")])),ggtheme = theme_minimal(base_size = 15))\n",
    "pdf(\"TMB_KM.pdf\",width = 5,height = 5)\n",
    "g\n",
    "dev.off()\n",
    "save(tmb,risk_tmb,x,km_tmb,file = \"TMB_coad.rdata\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebfb6b82-4794-4c2d-a272-d9814a4580f5",
   "metadata": {
    "tags": []
   },
   "source": [
    "### estimate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5563f696-19b1-4825-9fb0-608ac6b1cf9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(estimate)\n",
    "library(tidyverse)\n",
    "library(RColorBrewer)\n",
    "library(ggpubr)\n",
    "load(\"../estimate/TCGA/estimate.rdata\")\n",
    "load(\"../../Model_lasso_cox/模型构建/芯片建模_TCGA验证/OKModel_risk.rdata\")\n",
    "Expr = data.frame(entire_risk[,-c(17662:17670)]%>%t())\n",
    "Expr = data.frame(TCGA_exp_pd[,-c(1:10)]%>%t())\n",
    "write.table(Expr,file = \"TCGA_exp.txt\",row.names = T,quote = F,sep = \"\\t\")\n",
    "filterCommonGenes(input.f = \"TCGA_exp.txt\", \n",
    "                  output.f = \"TCGA_exp.gct\", \n",
    "                  id = \"GeneSymbol\")\n",
    "estimateScore(input.ds = \"TCGA_exp.gct\",\n",
    "              output.ds = \"TCGA_score_T.gct\", \n",
    "              platform = \"affymetrix\")#illumina;affymetrix\n",
    "scores <- read.table(\"TCGA_score_T.gct\",skip = 2,header = T)\n",
    "rownames(scores) <- scores[,1]\n",
    "scores <- t(scores[,3:ncol(scores)])\n",
    "scores\n",
    "TCGA_estimate = scores\n",
    "save(entire_estimate,TCGA_estimate,file = \"estimate.rdata\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4f6c20c-9152-4dd7-a4d5-5812f303a480",
   "metadata": {},
   "source": [
    "### ssgsea"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aba9a226-8d58-46e0-baa8-24148268476b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "library(GSVA)\n",
    "library(tidyverse)\n",
    "library(data.table)\n",
    "library(tidyr)\n",
    "library(pheatmap)\n",
    "library(RColorBrewer)\n",
    "library(ggpubr)\n",
    "load(\"Chip2_pd_exp for lasso_cox.rdata\")\n",
    "load(\"OKModel_risk.rdata\")\n",
    "entire_ = entire[,-c(17662:17668)]%>%t()\n",
    "TCGA_exp_ = TCGA_exp_pd[,-c(1:10)]%>%t()\n",
    "read.csv(\"29.csv\") %>%gather(set,gene)%>%unique()->x\n",
    "x = x[x$gene != \"\",]\n",
    "geneset29 = split(x$gene,x$set)\n",
    "ssGSEA_matrix <- gsva(expr = as.matrix(TCGA_exp_), \n",
    "                      gset.idx.list = geneset29,\n",
    "                      method = 'ssgsea',kcdf = 'Gaussian',abs.ranking = TRUE)\n",
    "scale_gsva_matrix <- t(scale(t(ssGSEA_matrix)))\n",
    "scale_gsva_matrix[scale_gsva_matrix< -2] <- -2\n",
    "scale_gsva_matrix[scale_gsva_matrix>2] <- 2\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe04b95e-8d6b-43fd-ba90-e5c5260c5c55",
   "metadata": {},
   "source": [
    "### GSVA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53769651-ab0c-4ca6-bd74-5f3266c0bcaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(tidyverse)\n",
    "library(clusterProfiler)\n",
    "library(msigdbr)  \n",
    "library(GSVA) \n",
    "library(GSEABase)\n",
    "library(pheatmap)\n",
    "library(limma)\n",
    "library(BiocParallel)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d694ba0d-0a05-43fb-bbf9-625b15e71b8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "KEGG_df_all <-  msigdbr(species = \"Homo sapiens\", # Homo sapiens or Mus musculus\n",
    "                        category = \"C2\",\n",
    "                        subcategory = \"CP:KEGG\") \n",
    "KEGG_df <- dplyr::select(KEGG_df_all,gs_name,gs_exact_source,gene_symbol)\n",
    "kegg_list <- split(KEGG_df$gene_symbol, KEGG_df$gs_name) \n",
    "\n",
    "\n",
    "GO_df_all <- msigdbr(species = \"Homo sapiens\",\n",
    "                     category = \"C5\")  \n",
    "GO_df <- dplyr::select(GO_df_all, gs_name, gene_symbol, gs_exact_source, gs_subcat)\n",
    "GO_df <- GO_df[GO_df$gs_subcat!=\"HPO\",]\n",
    "go_list <- split(GO_df$gene_symbol, GO_df$gs_name)\n",
    "\n",
    "Hallmark <-  msigdbr(species = \"Homo sapiens\", # Homo sapiens or Mus musculus\n",
    "                        category = \"H\") \n",
    "Hallmark_df <- dplyr::select(Hallmark,gs_name,gs_exact_source,gene_symbol)\n",
    "Hallmarklist <- split(Hallmark_df$gene_symbol, Hallmark_df$gs_name) ##按照gs_name给gene_symbol分组\n",
    "\n",
    "load(\"OKModel_risk.rdata\")\n",
    "dat = as.matrix(t(entire_risk[,1:17661]))\n",
    "geneset <- Hallmarklist#（kegg_list，go_list）\n",
    "gsva_mat <- gsva(expr=dat, \n",
    "               gset.idx.list=geneset, \n",
    "               kcdf=\"Gaussian\" ,#\"Gaussian\" for logCPM,logRPKM,logTPM, \"Poisson\" for counts\n",
    "               verbose=T, \n",
    "               parallel.sz = parallel::detectCores())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c603abc-0ebb-4de0-bac7-7177c6b987cc",
   "metadata": {},
   "source": [
    "### heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "dd77afab-f1b2-4de6-8986-4f52a2d169b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsva_mat = read.csv(\"gsva_go_matrix.csv\",header = T, row.names = 1)\n",
    "exp=\"high\"\n",
    "ctr=\"low\"\n",
    "design <- model.matrix(~0+factor(group_list$group))\n",
    "colnames(design) <- levels(factor(group_list$group))\n",
    "rownames(design) <- colnames(gsva_mat)\n",
    "contrast.matrix <- makeContrasts(contrasts=paste0(exp,'-',ctr),  #\"exp/ctrl\"\n",
    "                                 levels = design)\n",
    "fit1 <- lmFit(gsva_mat,design)                 \n",
    "fit2 <- contrasts.fit(fit1, contrast.matrix) \n",
    "efit <- eBayes(fit2)                         \n",
    "summary(decideTests(efit,lfc=0, p.value=0.05)) \n",
    "tempOutput <- topTable(efit, coef=paste0(exp,'-',ctr), n=Inf)\n",
    "degs <- na.omit(tempOutput) \n",
    "write.csv(degs,\"gsva_DDR_degs.results.csv\")\n",
    "degs =  read.csv(\"gsva_go_degs.results.csv\", header = T, row.names = 1)\n",
    "UP = rownames(degs[order(degs$logFC,decreasing = T),])[1:10];DOWN = rownames(degs[order(degs$logFC,decreasing = F),])[1:10]\n",
    "keep = c(UP,DOWN)\n",
    "padj_cutoff=0.05\n",
    "log2FC_cutoff=log2(1)\n",
    "keep <- rownames(degs[degs$adj.P.Val < padj_cutoff & abs(degs$logFC)>log2FC_cutoff, ])\n",
    "length(keep)\n",
    "keep = c(UP,DOWN)\n",
    "dat <- gsva_mat[keep,]\n",
    "dat = dat[,order(group_list$group)]\n",
    "annotation_row = data.frame(row.names = rownames(dat),anno = substr(rownames(dat),1,30))\n",
    "annotation_row$... = \"...\"\n",
    "annotation_row$anno = str_c(annotation_row$anno ,annotation_row$...)\n",
    "annotation_row$anno = gsub(\"_\",\" \",annotation_row$anno)\n",
    "groupcolor = c(\"#9d4351\",\"#6d5d80\");names(groupcolor) = c(\"high\",\"low\")\n",
    "ann_colors = list(group = groupcolor)\n",
    "pheatmap::pheatmap(dat, \n",
    "                   fontsize_row = 8,\n",
    "                   cluster_rows = F,\n",
    "                   cluster_cols = F,\n",
    "                   height = 5,\n",
    "                   width=7,\n",
    "                   scale = \"row\",\n",
    "                   annotation_col = group_list,\n",
    "                   annotation_colors = ann_colors,\n",
    "                   show_colnames = F,\n",
    "                   show_rownames = T,\n",
    "                   labels_row  = annotation_row$anno,\n",
    "                   color=colorRampPalette(c(\"navy\",\"white\",\"firebrick3\"))(100),\n",
    "                  # filename = paste0('GSVA_go_heatmap.pdf')\n",
    "                   filename = paste0('GSVA_kegg_heatmap.pdf')\n",
    "                  )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "e39eae39-009e-412f-8a62-efd89f28d748",
   "metadata": {},
   "outputs": [],
   "source": [
    "### DDR related pathway in hallmarks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 458,
   "id": "5e9ccea7-ef3a-4a7f-a7c0-631dd43daa3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "DDR_set = read.table(\"clipboard\",header = T,sep = \"\\t\")\n",
    "DDR_set$Pathway = gsub(\"\\\\s*\\\\(.*\",\"\",DDR_set$Pathway)\n",
    "DDR_set$Genes = gsub(\"\\\\s*\",\"\",DDR_set$Genes)\n",
    "DDR_set$Genes = strsplit(DDR_set$Genes,\",\")\n",
    "DDR_df = data.frame()\n",
    "for(i in 1:8){\n",
    "    \n",
    "df= data.frame(set  = DDR_set[i,1],genes = unlist(DDR_set[i,2]))\n",
    "DDR_df = rbind(DDR_df,df)    \n",
    "}\n",
    "DDR_list = split(DDR_df$genes,DDR_df$set)\n",
    "DDR_list[[\"NER_\"]] = DDR_list[[\"NER\"]][2:length(DDR_list[[\"NER\"]])]\n",
    "DDR_list_ = DDR_list[6]\n",
    "geneset <- DDR_list\n",
    "\n",
    "gsva_mat <- gsva(expr=dat, \n",
    "               gset.idx.list=geneset, \n",
    "               kcdf=\"Gaussian\" ,#\"Gaussian\" for logCPM,logRPKM,logTPM, \"Poisson\" for counts\n",
    "               verbose=T, \n",
    "               parallel.sz = parallel::detectCores())#调用所有核\n",
    "library(RColorBrewer)\n",
    "library(tidyr)\n",
    "library(ggpubr)\n",
    "data = data.frame(t(gsva_mat))\n",
    "data$ID = rownames(data);\n",
    "group_list$ID = rownames(group_list)\n",
    "data_ = merge(data,group_list,by = \"ID\")\n",
    "data_ = gather(data_,sets,values, -c(group,ID))\n",
    "colnames(data_)[2] = \"risk\"\n",
    "my_comparisons = list(c(\"high\",\"low\"))\n",
    "mypalette <- colorRampPalette(brewer.pal(8,\"Set1\"))\n",
    "g3 = ggplot(data_,aes(sets,values,fill = risk)) + \n",
    "  geom_boxplot(outlier.shape = 21,color = \"black\") + \n",
    "  theme_bw() + \n",
    "  labs(x = \"\", y = \"Gene expression\") +\n",
    "  theme(legend.position = \"top\") + \n",
    "  #theme(axis.text.x = element_text(angle=90,vjust = 0.7, hjust = 1))+\n",
    "  scale_fill_manual(values = mypalette(22)[c(1,6)])+ \n",
    "stat_compare_means(aes(group = risk,label = ..p.signif..),method = \"wilcox.test\")+\n",
    "theme(text =element_text(size = 20))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75b38850-576c-49dd-99db-728db090e6a1",
   "metadata": {},
   "source": [
    "### correlation analysis among score and DDR pathway"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8d33e89-fca9-40c3-b7ae-d9d82c61d92a",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(ggExtra)\n",
    "theme_global <- theme_bw()+theme(panel.grid.major = element_blank(),panel.grid.minor=element_blank()) + theme(aspect.ratio = 1)+  theme(\n",
    "    text = element_text(size = 15),  # 调整所有文本的字体大小\n",
    "    axis.text = element_text(size = 15),  # 调整坐标轴标签的字体大小\n",
    "    axis.title = element_text(size = 15),  # 调整坐标轴标题的字体大小\n",
    "    legend.title = element_text(size = 14),  # 调整图例标题的字体大小\n",
    "    legend.text = element_text(size = 12)  # 调整图例文本的字体大小\n",
    "  )\n",
    "risk_colors <- c(\"low\" = \"#469f6f\", \"high\" = \"#e41a1c\") \n",
    "data = read.csv(\"gsva_Hallmark_matrix.csv\",header = T,row.names = 1)\n",
    "data = as.data.frame(t(data))\n",
    "data_ = data[,c(12,35,14,18)]\n",
    "data_$ID = rownames(data_)\n",
    "df = data.frame(ID = rownames(entire_risk),riskScore = entire_risk[,17669],risk =entire_risk[,17670])\n",
    "data__ =merge(df,data_, by = \"ID\")\n",
    "\n",
    "colnames(data__)[4:7] = gsub(\"_\",\" \",substring(colnames(data__)[4:7],10))\n",
    "colnames(data__)[6] = \"EMT\"\n",
    "pp = list()\n",
    "for(i in 4:ncol(data__)){\n",
    "g1 = ggscatter(data__, x =colnames(data__)[2] ,y = colnames(data__)[i],\n",
    "          add = \"reg.line\", conf.int = T,\n",
    "          color = \"risk\",\n",
    "          add.params = list(fill = \"lightgray\",color = \"black\"),\n",
    "          ggtheme = theme_minimal(), margin.plot =  \"density\"\n",
    ")+ stat_cor(method = \"pearson\",\n",
    "           label.x = median(data__[,2]), label.y =max(data__[,i]),color='red',size =7)+theme_global+\n",
    "    scale_color_manual(values =risk_colors ,guide = F)\n",
    "g2 = ggMarginal(g1, type=\"density\",xparams = list( fill=\"#3c69ac\", size = 0.5,alpha=1), yparams = list(fill=\"#bb4962\", size = 0.5,alpha=1))\n",
    "pp[[i]] = g2\n",
    "}\n",
    "cor_4_hallmark = data__\n",
    "save(cor_4_hallmark,file = \"cor_4_hallmark.rdata\")\n",
    "plot_ = pp[4:7]\n",
    "pdf(\"cor_list.pdf\",width = 20,height = 5)\n",
    "cowplot::plot_grid(plotlist = plot_, align = \"h\", nrow = 1)\n",
    "dev.off()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
